src/core/net/ip6.cpp

```diff
@@ -1007,45 +1005,62 @@ const NetifUnicastAddress *Ip6::SelectSourceAddress(MessageInfo &aMessageInfo)
                 // Rule 2: Prefer appropriate scope
                 if (addr->GetScope() >= overrideScope)
                 {
-                    rvalAddr = addr;
-                    rvalIface = candidateId;
-                    rvalPrefixMatched = candidatePrefixMatched;
+                    goto replace;
```

Comment:
We can avoid the `goto` by breaking out the body of this for-loop into an inline function like:

```c++
bool IsSourceAddressPreferred(const Address &aDestination, const NetifUnicastAddress &aSource, const NetifUnicastAddress &aCandidate)
```

Thoughts?